src/osp/scientific/kepler.cpp

```diff
@@ -0,0 +1,480 @@
+ */
+static double wrap_angle(double const angle)
+{
+    double wrapped = fmod(angle, 2.0 * PI);
+    if (wrapped < 0)
+    {
+        wrapped += 2.0 * PI;
+    }
+    return wrapped;
+}
+
+/**
+ * @brief Solve Kepler's equation for elliptic orbits (E = M + e * sin(E))
+ * 
+ * @return Eccentric anomaly E
+ */
+static double solve_kepler_elliptic(double const meanAnomaly, double const eccentricity, int steps)
+{
+    double const tolerance = 1.0E-10;
+
+    // If eccentricity is low, then this is a circular orbit, so M = E
+    if (std::abs(eccentricity) <= 1.0E-8)
+    {
+        return meanAnomaly;
+    }
+
+    double delta;
+
+    // initial guess
+    double E = meanAnomaly;
+
+    // newton-rhapson
+    do
+    {
+        double f_E = E - eccentricity * sin(E) - meanAnomaly;
+        double df_E = 1.0 - eccentricity * cos(E);
+
+        delta = f_E / df_E;
+        E = E - delta;
+
+        E = wrap_angle(E); // Sometimes this sometimes fails to converge without this line
+        --steps;
+    } while (std::abs(delta) > tolerance && steps > 1);
+    return E;
+}
+
+/**
+ * @brief Solve Kepler's equation for hyperbolic orbits (E = M + e * sinh(E))
+ * 
+ * @return Eccentric anomaly E
+ */
+static double solve_kepler_hyperbolic(double const meanAnomaly, double const eccentricity, int steps)
+{
+    double const tolerance = 1.0E-10;
+
+    double delta;
+
+    // initial guess
+    double E = std::log(std::abs(meanAnomaly) + 0.1);
+
+    // newton-rhapson
+    do
+    {
+        double f_E = eccentricity * sinh(E) - E - meanAnomaly;
+        double df_E = eccentricity * cosh(E) - 1.0;
+
+        delta = f_E / df_E;
+        E = E - delta;
+        --steps;
+    } while (std::abs(delta) > tolerance && steps > 1);
+    return E;
+}
+
+/**
+ * @brief Convert eccentric anomaly (E) to true anomaly (ν) for elliptic orbits
+ */
+static double elliptic_eccentric_anomaly_to_true_anomaly(double const &ecc, double const &E)
+{
+    // This formula supposedly has better numerical stability than the one below
+    double const beta = ecc / (1 + std::sqrt(1 - ecc * ecc));
+    return E + 2.0 * atan2(beta * sin(E), 1 - beta * cos(E));
+
+    // return 2.0 * atan2(std::sqrt(1 + ecc) * sin(E / 2.0), std::sqrt(1 - ecc) * cos(E / 2.0));
+}
+
+/**
+ * @brief Convert hyperbolic eccentric anomaly (E) to true anomaly (ν)
+ */
+static double hyperbolic_eccentric_anomaly_to_true_anomaly(double const &ecc, double const &H)
+{
+    return 2 * atan(sqrt((ecc + 1) / (ecc - 1)) * tanh(H / 2));
+}
+
+/**
+ * @brief Convert true anomaly (ν) to eccentric anomaly (E) for elliptic orbits
+ */
+static double true_anomaly_to_elliptic_eccentric_anomaly(double const &ecc, double const &v)
+{
+    return 2.0 * atan(std::sqrt((1 - ecc) / (1 + ecc)) * tan(v / 2.0));
+}
+
+/**
+ * @brief Convert true anomaly (ν) to hyperbolic eccentric anomaly (E)
+ */ 
+static double true_anomaly_to_hyperbolic_eccentric_anomaly(double const &ecc, double const &v)
```

Comment:
Is there any value in taking `ecc` and `v` as references instead of by value?

At least on x86_64 linux, with clang, `sizeof(double)` is 8bytes, which is the same as a pointer.

So on most platforms, when passing by value, the double should be passed via register (assuming few enough parameters). On platforms where there isnt enough space to pass by register, it'll implicitly pass by reference / pointer regardless.